File: Programmers/BFS/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<bool> visited;

bool canChange(string cur, string target) {
    int cnt = 0;
    for (int i = 0; i < cur.size(); i++) {
        if (cur[i] != target[i]) cnt++;
    }
    return cnt == 1;
}

int BFS(string begin, string target, vector<string> words) {
    queue<pair<string, int>> q;
    q.push({begin, 0});

    while(!q.empty()) {
        auto [cur, cnt] = q.front();
        q.pop();

        if (cur == target) return cnt;

        for (int i = 0; i < words.size(); i++) {
            if (!visited[i] && canChange(cur, words[i])) {
                q.push({words[i], cnt + 1});
                visited[i] = true;
            }
        }
    }
    return 0;
}

int solution(string begin, string target, vector<string> words) {
    visited.assign(words.size(), false);

    return BFS(begin, target, words);
}

File: Programmers/BFS/단어 변환.py
from collections import deque

def canChange(cur, target):
    cnt = sum([1 for a, b in zip(cur, target) if a != b])
    return cnt == 1

def bfs(begin, target, words):
    queue = deque()
    n = len(words)
    queue.append((begin, 0))
    visited = [False] * n

    while queue:
        cur, cnt = queue.popleft()

        if cur == target: return cnt

        for i in range(n):
            if not visited[i] and canChange(cur, words[i]):
                visited[i] = True
                queue.append((words[i], cnt + 1))
    return 0

def solution(begin, target, words):
    return bfs(begin, target, words)